datanode/gateway/middleware.go

```diff
@@ -124,3 +132,92 @@ func (i *InjectableResponseWriter) Write(data []byte) (int, error) {
 func (i *InjectableResponseWriter) SetHeaders(headers http.Header) {
 	i.headers = headers
 }
+
+type SubscriptionRateLimiter struct {
+	log *logging.Logger
+	m   map[string]uint32
+	mu  sync.Mutex
+
+	MaxSubscriptions uint32
+}
+
+func NewSubscriptionRateLimiter(
+	log *logging.Logger,
+	maxSubscriptions uint32,
+) *SubscriptionRateLimiter {
+	return &SubscriptionRateLimiter{
+		log:              log,
+		MaxSubscriptions: maxSubscriptions,
+		m:                map[string]uint32{},
+	}
+}
+
+func (s *SubscriptionRateLimiter) Inc(ip string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	cnt := s.m[ip]
+	if cnt == s.MaxSubscriptions {
+		return ErrMaxSubscriptionReached(ip, s.MaxSubscriptions)
+	}
+	s.m[ip] = cnt + 1
+	return nil
+}
+
+func (s *SubscriptionRateLimiter) Dec(ip string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	cnt := s.m[ip]
+	s.m[ip] = cnt - 1
+}
+
+func (s *SubscriptionRateLimiter) WithSubscriptionRateLimiter(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// is that a subscription?
+		if _, ok := w.(http.Hijacker); !ok {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		if ip, err := getIP(r); err != nil {
+			s.log.Debug("couldn't get client ip", logging.Error(err))
+		} else {
+			if err := s.Inc(ip); err != nil {
+				s.log.Error("couldn't client reached max sub", logging.Error(err))
```

Comment:
This could be a bit clearer

Reply:
haha I don't what I was trying to say here to be fair.

Reply:
ddressed